apps/slack/slack_app.py

```diff
@@ -0,0 +1,154 @@
+from django.conf import settings
+from django.contrib.messages import constants
+from django.urls import reverse
+from django.utils.functional import SimpleLazyObject
+from slack_bolt import App, BoltRequest, BoltResponse
+from slack_bolt.adapter.django import SlackRequestHandler
+from slack_bolt.logger import get_bolt_logger
+from slack_bolt.oauth import OAuthFlow
+from slack_bolt.oauth.callback_options import CallbackOptions, FailureArgs, SuccessArgs
+from slack_bolt.oauth.internals import build_detailed_error
+from slack_bolt.oauth.oauth_settings import OAuthSettings
+from slack_bolt.util.utils import create_web_client
+from slack_sdk.oauth import OAuthStateUtils
+from slack_sdk.oauth.installation_store import Installation
+
+from apps.service_providers.const import MESSAGING
+from apps.service_providers.models import MessagingProvider, MessagingProviderType
+from apps.web.meta import absolute_url
+
+from . import const
+from .const import INSTALLATION_CONFIG
+from .slack_datastores import DjangoInstallationStore, DjangoOAuthStateStore
+
+bolt_logger = get_bolt_logger(App)
+
+
+def get_slack_app():
+    client_id, client_secret, signing_secret, scopes = (
```

Comment:
Should these not come from the slack messaging provider instance?

Reply:
Interesting question but I don't think so. This app represents the slack app which you can then add to different slack workspaces. Adding it to a workspace is the 'messaging provider' thing but each OCS instance should only have one slack app (I think).

But I'll give it some more thought.

Reply:
I did some more thinking and I'm sure this is the correct way to do it. If we make it a slack app per provider then users will need to create lots of different slack apps and then add them to their workspaces which is unnecessary. The message provider is connecting the Slack app to a slack workspace and to an OCS team which allows it to interact with the workspace.

I think it would be the same for any oauth based app. E.g. with CommCare we'd create one oauth app in CommCare for the Dimagi OCS instance but users would need to use oath to connect their domain.